EIPS/eip-7701.md

```diff
@@ -40,382 +43,121 @@ Transactions of this type are referred to as "AA transactions".
-| `role_sender_deployment`, `role_sender_validation`, `role_paymaster_validation` | All tx parameters, `current_frame_role`, `tx_hash_for_signature`                                           |
-| `role_paymaster_post_op`                                                        | All tx parameters, `current_frame_role`, `tx_hash_for_signature`, `execution_status`, `execution_gas_cost` |
-| `role_sender_execution`                                                         | Only `current_frame_role`                                                                                  |
-
-
-Contact should first call `TXPARAMLOAD 0xF0` (`current_frame_role`) to determine the current frame role.
-
-In case `current_frame_role` is not set for the current frame it has a default value of `0`.
-
-### `TXPARAM*` opcodes gas prices
-
-The new opcodes have gas prices equivalent to corresponding `CALLDATA*` opcodes:
-
-1. `TXPARAMLOAD` - static value of `3` gas
-2. `TXPARAMSIZE`- static value of `2` gas
-3. `TXPARAMCOPY` - dynamic formula
-
-    ```
-    minimum_word_size = (size + 31) / 32
-    static_gas = 3
-    dynamic_gas = 3 * minimum_word_size + memory_expansion_cost
-    ```
-
-#### Time range validity (optional)
-
-If the `valid_until` field is non-zero, the transaction is only valid for inclusion in a block with a timestamp at most `valid_until` value.
-Similarly, the transaction is only valid for inclusion in blocks with a timestamp at most the `valid_after` value.
-
-#### Sender Deployment (optional)
-
-Inputs to the `deployer` contract are not defined by the protocol and are controlled by the `deployer_data` parameter.
-
-The sender deployment frame MUST result in the `sender` address becoming initialized with contract code.
-
-This step is performed with the `role_sender_deployment` role.
-
-#### Sender Validation
+### `CURRENT_ROLE` and `ACCEPT_ROLE` opcodes
 
-This step is performed with the `role_sender_validation` role.
+`current_frame_role` is a context variable set by the AA transaction to the current role.
+During AA transactions it is set to the current role in the transaction's lifecycle for each top level call. 
+During non-AA transactions it is always set to `ROLE_SENDER_EXECUTION`.
+It remains unchanged on `DELEGATECALL` but is reset to `ROLE_SENDER_EXECUTION` on `CALL` / `STATICCALL` / `CALLCODE`. This behavior resembles `msg.sender`.
 
-In order for the transaction to be considered valid, the
-sender validation frame MUST return without reverting.
+The `CURRENT_ROLE` opcode returns the `current_frame_role` value.
 
-#### Paymaster Validation (optional)
+The `ACCEPT_ROLE` opcode is equivalent to `RETURN` in the sense that it copies a memory slice, ends execution, and pastes the memory slice onto parent returndata and parent memory, and except:
 
-This step is performed with the `role_paymaster_validation` role.
+* It accepts the `frame_role` as the additional input parameter, and reverts if it differs from the `current_frame_role`
```

Comment:
```suggestion
* It accepts the `frame_role` as the additional input parameter, and reverts if it differs from the `current_context_role`
```

Reply:
@drortirosh does `current_context_role` represent the global role (based on the current stage in the transaction), or did you just rename `current_frame_role` but left its meaning unchanged?  I.e. inside a CALL, is `current_context_role` always set to ROLE_EXECUTION regardless of global role, as before?